File: 024.SwapNodesInPairs/wyrj/SwapNodesInPairs.js
/* 148ms */
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function(head) {
    var retHead = head, prev = null, curr = head;
    while(curr && curr.next) {
        if (!prev) {
            retHead = curr.next;
            curr.next = retHead.next;
            retHead.next = curr;
        } else {
            prev.next = curr.next;
            curr.next = prev.next.next;
            prev.next.next = curr;
        }
        prev = curr;
        curr = curr.next;
    }
    return retHead;
};

File: 024.SwapNodesInPairs/wyrj/SwapNodesInPairs.c
/* 0ms */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* swapPairs(struct ListNode* head) {
    struct ListNode *retHead = head, *prev = NULL, *curr = head;
    while(NULL != curr && NULL != curr->next) {
        if (NULL == prev) {
            retHead = curr->next;
            curr->next = retHead->next;
            retHead->next = curr;
        } else {
            prev->next = curr->next;
            curr->next = prev->next->next;
            prev->next->next = curr;
        }
        prev = curr;
        curr = curr->next;
    }
    return retHead;
}